### PR TITLE
Speed up Reranking tasks

### PR DESCRIPTION
**Pull request by AlexeyVatolin**
Pull request to solve issue https://github.com/embeddings-benchmark/mteb/issues/738

## Checklist for adding MMTEB dataset

Reason for dataset addition:

- [x] I have tested that the dataset runs with the `mteb` package.
- [x] I have run the following models on the task (adding the results to the pr). These can be run using the `mteb -m {model_name} -t {task_name}` command.
  - [x] `sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2`
  - [x] `intfloat/multilingual-e5-small`
- [x] I have checked that the performance is neither trivial (both models gain close to perfect scores) nor random (both models gain close to random scores).
- [x] If the dataset is too big (e.g. >2048 examples), considering using `self.stratified_subsampling() under dataset_transform()`
- [x] I have filled out the metadata object in the dataset file (find documentation on it [here](https://github.com/embeddings-benchmark/mteb/blob/main/docs/adding_a_dataset.md#2-creating-the-metadata-object)).
- [x] Run tests locally to make sure nothing is broken using `make test`. 
- [x] Run the formatter to format the code using `make lint`. 
- [x] I have added points for my submission to the [points folder](https://github.com/embeddings-benchmark/mteb/blob/main/docs/mmteb/points.md) using the PR number as the filename (e.g. `438.jsonl`).

### CODE REVIEW

File: mteb/evaluation/evaluators/RerankingEvaluator.py [hunk 1]
```diff
@@ -188,6 +186,19 @@ def compute_metrics_individual(self, model):
 
         return {"map": mean_ap, "mrr": mean_mrr}
 
+    def _encode_unqiue_texts(self, all_texts, encode_queries_func):
```
Comment (Ruqyai) [inline mteb/evaluation/evaluators/RerankingEvaluator.py:189]: Thank you for the useful changes! Adding the `_encode_unqiue_texts `method to improve encoding efficiency is a great idea. I have a few minor suggestion:
Correct the typo in the name from `_encode_unqiue_texts` to `_encode_unique_texts`.

File: docs/mmteb/points/792.jsonl [hunk 2]
```diff
@@ -0,0 +1 @@
+{"GitHub": "AlexeyVatolin", "Bug fixes": 2}
```
Comment (KennethEnevoldsen) [inline docs/mmteb/points/792.jsonl:1]:
```suggestion
{"GitHub": "AlexeyVatolin", "Bug fixes": 4}
{"GitHub": "KennethEnevoldsen", "PR Review": 2}
{"GitHub": "Ruqyai", "PR Review": 2}

```

**Review by KennethEnevoldsen — APPROVED** [review]
This looks good. Will you do a check that the scores are still the same when running a reranking task?

**Comment by AlexeyVatolin** [comment]
@KennethEnevoldsen, I checked for `intfloat/multilingual-e5-small` and `entence-transformers/paraphrase-multilingual-MiniLM-L12-v2`, all ranking scores match to 3 decimal places. Do I need to post the results of the models? 

I also checked if my code slows down other ranking tasks (except MindSmallReranking) using `evaluation_time` from the json of the evaluation results. The maximum slowdown is 0.62 seconds and the total speedup is 915 seconds for the `intfloat/multilingual-e5-small` model.

| model name                                                  | task name                 |     speedup |
|:------------------------------------------------------------|:--------------------------|------------:|
| sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2 | MindSmallReranking        | 824.25      |
| intfloat/multilingual-e5-small                              | MindSmallReranking        | 784.067     |
| intfloat/multilingual-e5-small                              | AlloprofReranking         |  88.8254    |
| sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2 | AlloprofReranking         |  30.8255    |
| intfloat/multilingual-e5-small                              | CMedQAv1-reranking        |  19.9761    |
| sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2 | CMedQAv1-reranking        |  16.0283    |
| intfloat/multilingual-e5-small                              | CMedQAv2-reranking        |  13.249     |
| sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2 | CMedQAv2-reranking        |  11.6278    |
| intfloat/multilingual-e5-small                              | SciDocsRR                 |   5.27587   |
| sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2 | SciDocsRR                 |   5.20219   |
| intfloat/multilingual-e5-small                              | SyntecReranking           |   1.63771   |
| intfloat/multilingual-e5-small                              | MMarcoReranking           |   1.26827   |
| sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2 | MMarcoReranking           |   0.378537  |
| sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2 | SyntecReranking           |   0.351005  |
| intfloat/multilingual-e5-small                              | T2Reranking               |   0.321867  |
| sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2 | StackOverflowDupQuestions |   0.293564  |
| intfloat/multilingual-e5-small                              | StackOverflowDupQuestions |   0.28991   |
| intfloat/multilingual-e5-small                              | MIRACLReranking           |   0.1643    |
| sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2 | AskUbuntuDupQuestions     |   0.0437329 |
| intfloat/multilingual-e5-small                              | AskUbuntuDupQuestions     |  -0.0347993 |
| sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2 | MIRACLReranking           |  -0.0939496 |
| sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2 | T2Reranking               |  -0.620706  |

**Comment by KennethEnevoldsen** [comment]
> @KennethEnevoldsen, I checked for intfloat/multilingual-e5-small and entence-transformers/paraphrase-multilingual-MiniLM-L12-v2, all ranking scores match to 3 decimal places. Do I need to post the results of the models?

All good! just wanted the confirmation

Thanks for the speed-up as well this looks very reasonable


we might consider adding this change to retrieval tasks as well

**Comment by Muennighoff** [comment]
> @KennethEnevoldsen, I checked for `intfloat/multilingual-e5-small` and `entence-transformers/paraphrase-multilingual-MiniLM-L12-v2`, all ranking scores match to 3 decimal places. Do I need to post the results of the models?
> 
> I also checked if my code slows down other ranking tasks (except MindSmallReranking) using `evaluation_time` from the json of the evaluation results. The maximum slowdown is 0.62 seconds and the total speedup is 915 seconds for the `intfloat/multilingual-e5-small` model.
> 
> model name	task name	speedup
> sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2	MindSmallReranking	824.25
> intfloat/multilingual-e5-small	MindSmallReranking	784.067
> intfloat/multilingual-e5-small	AlloprofReranking	88.8254
> sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2	AlloprofReranking	30.8255
> intfloat/multilingual-e5-small	CMedQAv1-reranking	19.9761
> sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2	CMedQAv1-reranking	16.0283
> intfloat/multilingual-e5-small	CMedQAv2-reranking	13.249
> sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2	CMedQAv2-reranking	11.6278
> intfloat/multilingual-e5-small	SciDocsRR	5.27587
> sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2	SciDocsRR	5.20219
> intfloat/multilingual-e5-small	SyntecReranking	1.63771
> intfloat/multilingual-e5-small	MMarcoReranking	1.26827
> sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2	MMarcoReranking	0.378537
> sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2	SyntecReranking	0.351005
> intfloat/multilingual-e5-small	T2Reranking	0.321867
> sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2	StackOverflowDupQuestions	0.293564
> intfloat/multilingual-e5-small	StackOverflowDupQuestions	0.28991
> intfloat/multilingual-e5-small	MIRACLReranking	0.1643
> sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2	AskUbuntuDupQuestions	0.0437329
> intfloat/multilingual-e5-small	AskUbuntuDupQuestions	-0.0347993
> sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2	MIRACLReranking	-0.0939496
> sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2	T2Reranking	-0.620706

Looks great! If it's the same down to 3 decimal places, it seems fine to merge this without backward compat

**Comment by AlexeyVatolin** [comment]
@KennethEnevoldsen, points in `main` branch are also broken, but I fixed it in my last commit.
Also `test_dataset_availability` test in `main` are broken, here are logs
```
E           AssertionError: Datasets not available on Hugging Face:
E             iwslt2017 - revision c18a4f81a47ae6fa079fe9d32db288ddde38451d
E             myanmar_news - revision b899ec06227db3679b0fe3c4188a6b48cc0b65eb
E             moroco - revision d64d9b8cd876056a5c24552afe3caf7e6fd26c8e
E             ro_sent - revision 155048684cea7a6d6af1ddbfeb9a04820311ce93
E             swedish_reviews - revision 105ba6b3cb99b9fd64880215be469d60ebf44a1b
E             indonli - revision 3c976110fc13596004dc36279fc4c453ff2c18aa
E             ro_sts - revision 41a33183b739070f3d46d9d446492c1d2f98ce1a
E           assert False
```

**Comment by KennethEnevoldsen** [comment]
> @KennethEnevoldsen, points in main branch are also broken, but I fixed it in my last commit.
Also test_dataset_availability test in main are broken, here are logs

pulling from main should fix the issue

**Comment by AlexeyVatolin** [comment]
@KennethEnevoldsen, all the tests are working now. Can we merge this pull request?

**Comment by KennethEnevoldsen** [comment]
Yep @AlexeyVatolin. If you wish I could imagine a similar function could be added to retrieval tasks

**Comment by Muennighoff** [comment]
Amazing work on this! Wouldn't it be cleaner to just remove the duplicates and update the dataset at https://huggingface.co/datasets/mteb/mind_small? Then we just update the revision here in the code and we don't need the extra complexity in the eval code to deduplicate. What do you think? 🤔

I think that always getting
```
WARNING:mteb.evaluation.evaluators.RerankingEvaluator:A total on 2652814/2658091 duplicate texts were found during encoding.
 Only encoding unique text and duplicating embeddings across.
```
is annoying 😁

**Comment by KennethEnevoldsen** [comment]
Deduplicating would def. be better. Though that would probably change the scores.

**Comment by Muennighoff** [comment]
Why would it change the scores? Wouldn't it lead to the same dataset to be encoded as the solution here? I'm probably missing something obvious 😅

**Comment by KennethEnevoldsen** [comment]
I was imagining that the queries might not have annotation for all duplicates (which is a mistake in annotation), but that would be my guess. {q1: {c1: 1}, q2: {c2: 1}}, but c2 == c1 so it should really be: {q1: {c1: 1, c2: 1}, q2: {c2: 1, c1: 1}}, however even that would be different from {q1: {c1_dedup: 1}, q2: {c1_dedup: 1}} for some metrics (or am I missing something here?)

**Comment by Muennighoff** [comment]
cc @AlexeyVatolin do you have thoughts on this? You probably know best 😁

**Pull request merged**